v2/pkg/protocols/common/protocolstate/headless.go

```diff
@@ -0,0 +1,77 @@
+package protocolstate
+
+import (
+	"strings"
+
+	"github.com/go-rod/rod"
+	"github.com/go-rod/rod/lib/proto"
+	"github.com/projectdiscovery/networkpolicy"
+	errorutil "github.com/projectdiscovery/utils/errors"
+	urlutil "github.com/projectdiscovery/utils/url"
+	"go.uber.org/multierr"
+)
+
+// initalize state of headless protocol
+
+var (
+	ErrURLDenied         = errorutil.NewWithFmt("headless: url %v dropped by rule: %v")
+	networkPolicy        *networkpolicy.NetworkPolicy
+	allowLocalFileAccess bool
+)
+
+// ValidateNFailRequest validates and fails request
+// if the request does not respect the rules, it will be canceled with reason
+func ValidateNFailRequest(page *rod.Page, e *proto.FetchRequestPaused) error {
+	reqURL := e.Request.URL
+	normalized := strings.ToLower(reqURL) // normalize url to lowercase
+	if !allowLocalFileAccess && strings.HasPrefix(normalized, "file:") {
```

Comment:
we can also disallow other protocols such as `chrome://` to avoid access to chrome settings